### PR TITLE
Build variable replacement dict on init

### PR DESCRIPTION
**Pull request by jgkamat**
Essentially a little more fancy version of http://www.github.com/qutebrowser/qutebrowser/pull/4536

Continuation of https://github.com/qutebrowser/qutebrowser/pull/4652#issuecomment-484351967 

Haven't run all tests on this yet, but it seemed to pass what I thought would break. I'll fix anything if it comes up though (still not 100% sure this is safe).

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4731)

### CODE REVIEW

File: qutebrowser/commands/runners.py [hunk 1]
```diff
@@ -33,6 +34,12 @@
 from qutebrowser.utils import message, objreg, qtutils, usertypes, utils
 from qutebrowser.misc import split, objects
 
+MYPY = False
+if MYPY:
+    # pylint: disable=unused-import
+    from qutebrowser.mainwindow import tabbedbrowser
+    # pylint: enable=unused-import
+ReplacementFunction = typing.Callable[['tabbedbrowser.TabbedBrowser'], str]
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]:
```suggestion
_ReplacementFunction = typing.Callable[['tabbedbrowser.TabbedBrowser'], str]
```

as it shouldn't be needed from outside of this module

File: qutebrowser/commands/runners.py [hunk 2]
```diff
@@ -59,58 +66,67 @@ def _current_url(tabbed_browser):
...
-    """Utility function to replace variables like {url} in a list of args."""
-    tabbed_browser = objreg.get('tabbed-browser', scope='window',
-                                window=win_id)
-    url = lambda: _current_url(tabbed_browser)
-
-    variables = {
-        'url': lambda: url().toString(
+def _init_variable_replacements() -> typing.Dict[str, ReplacementFunction]:
+    """Return a dict from variable replacements to fns processing them."""
+    url = _current_url
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: Could just rename the `_current_url` function to `_url` instead :wink:

File: qutebrowser/commands/runners.py [hunk 3]
```diff
@@ -59,58 +66,67 @@ def _current_url(tabbed_browser):
...
+    for key in list(replacements):
         modified_key = '{' + key + '}'
-        variables[modified_key] = lambda x=modified_key: x
+        # x = modified_key is to avoid binding x as a closure
+        replacements[modified_key] = (
+            lambda _, x=modified_key: x)  # type: ignore
+    return replacements
+
 
+VARIABLE_REPLACEMENTS = types.MappingProxyType(_init_variable_replacements())
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: Why the `MappingProxyType`?

Comment (user202729) [reply qutebrowser/commands/runners.py]: Likely to ensure that it is not accidentally changed in some other places.

Comment (The-Compiler) [reply qutebrowser/commands/runners.py]: `Mapping` is just a more loose version of `Dict` (like `Sequence` or `Iterable` compared to `List`), I don't think there's currently a way to make mypy check immutability.

As for using `types.MappingProxyType`: I don't think this makes sense here. The name clearly conveys that it's constant, so if it'd be modified somewhere else, that'd stand out anyways. That kind of stuff IMHO makes sense where accidental mutations can happen easily (like with objects coming from the config), but I don't think that's the case here.

Comment (user202729) [reply qutebrowser/commands/runners.py]: No, `MutableMapping` is a more loose version of `Dict`. `Mapping` is immutable.

Comment (user202729) [reply qutebrowser/commands/runners.py]: There's also `from typing_extensions import Final` (see https://github.com/python/mypy/pull/5522 for usage), but that is another kind of immutability - that the variable cannot be reassigned. It's not the case here.

**Comment by user202729** [comment]
Unfortunately, mypy doesn't type check dynamically typed functions. So using `Mapping` to enforce immutability at check time won't work.

**Comment by The-Compiler** [comment]
> No, `MutableMapping` is a more loose version of `Dict`. `Mapping` is immutable.

My bad, you're right!

> Unfortunately, mypy doesn't type check dynamically typed functions. So using `Mapping` to enforce immutability at check time won't work.

Since we're annotating more and more code in qutebrowser, I'd still go ahead with using `typing.Mapping` plus manual code review - the chance that someone accidentally mutates this dict seems pretty low anyways.

**Comment by The-Compiler** [comment]
Thanks!

**Pull request merged**